src/core/DY.Core.Printing.fst

```diff
@@ -0,0 +1,113 @@
+  
+  | Rand usage label len time -> Printf.sprintf "Nonce #%d" time
+  
+  | Concat (Literal s) right -> Printf.sprintf "%s%s" 
+                                  (uint_list_to_string (FStar.Seq.seq_to_list s)) (bytes_to_string right)
+  | Concat left right -> Printf.sprintf "[%s,%s]" (bytes_to_string left) (bytes_to_string right)
+  
+  | AeadEnc key nonce msg ad -> Printf.sprintf "AeadEnc(key=(%s), nonce=(%s), msg=(%s), ad=(%s))" 
+                              (bytes_to_string key) (bytes_to_string nonce) (bytes_to_string msg)
+                              (bytes_to_string ad)
+  
+  | Pk sk -> Printf.sprintf "Pk(sk=(%s))" (bytes_to_string sk)
+  | PkEnc pk nonce msg -> Printf.sprintf "PkEnc(pk=(%s), nonce=(%s), msg=(%s))" 
+                              (bytes_to_string pk) (bytes_to_string nonce) (bytes_to_string msg)
+  
+  | Vk sk -> Printf.sprintf "Public Key (%s)" (bytes_to_string sk)
+  | Sign sk nonce msg -> Printf.sprintf "sig(sk=(%s), nonce=(%s), msg=(%s))" 
+                              (bytes_to_string sk) (bytes_to_string nonce) (bytes_to_string msg)
+
+  | Hash msg -> Printf.sprintf "Hash(%s)" (bytes_to_string msg)
+
+  | DhPub sk -> Printf.sprintf "DhPub(sk=(%s))" (bytes_to_string sk)
+  | Dh sk pk -> Printf.sprintf "Dh(sk=(%s), pk=(%s))" (bytes_to_string sk) (bytes_to_string pk)
+
+val trace_event_to_string: trace_event -> nat -> string
+let trace_event_to_string tr_event i =
+  match tr_event with
+  | MsgSent msg -> (
+    Printf.sprintf "{\"TraceID\": %d, \"Type\": \"Message\", \"Content\": \"%s\"}\n"
+      i (bytes_to_string msg)
+  )
+  | RandGen usg lab len -> (
+    Printf.sprintf "{\"TraceID\": %d, \"Type\": \"Nonce\", \"Usage\": \"%s\", \"Label\": \"%s\"}\n" 
+    i (usage_to_string usg) (label_to_string lab)
+  )
+  | Corrupt prin sess_id -> ""
+  | SetState prin sess_id content -> (
+    Printf.sprintf "{\"TraceID\": %d, \"Type\": \"Session\", \"SessionID\": %d, \"Principal\": \"%s\", \"Content\": \"%s\"}\n"
+      i sess_id prin (bytes_to_string content)
+  )
+  | Event prin tag content -> (
+    Printf.sprintf "{\"TraceID\": %d, \"Type\": \"Event\", \"Principal\": \"%s\", \"Tag\": \"%s\", \"Content\": \"%s\"}\n" 
+      i prin tag (bytes_to_string content)
+  )
+
+val trace_to_string_helper: (tr:trace) -> (i:nat{i = length tr}) -> string
+let rec trace_to_string_helper tr i =
+  match tr with
+  | Nil -> ""
+  | Snoc ptr ev -> trace_to_string_helper ptr (i-1) ^ trace_event_to_string ev i
+
+/// This method can be used to print the trace of
+/// an example protocol run to the console.
+/// The output can be parsed and converted to a
+/// Plantuml sequence diagram with the following
+/// experimental Python script: https://github.com/fabian-hk/dolev-yao-star-output-parser
+///
+/// Usage:
+/// let* tr = get_trace in
+/// let _ = IO.debug_print_string (trace_to_string tr) in
+
+val trace_to_string: trace -> string
+let trace_to_string tr =
+  trace_to_string_helper tr (length tr)
```

Comment:
```suggestion
/// This method can be used to print the trace of
/// an example protocol run to the console.
/// The output can be parsed and converted to a
/// Plantuml sequence diagram with the following
/// experimental Python script: https://github.com/fabian-hk/dolev-yao-star-output-parser
///
/// Usage:
/// let* tr = get_trace in
/// let _ = IO.debug_print_string (trace_to_string tr) in

val trace_to_string: trace -> string
let rec trace_to_string tr =
  match tr with
  | Nil -> ""
  | Snoc ptr ev -> trace_to_string ptr ^ trace_event_to_string ev (length tr)
```

I think we can avoid the `i` and therefore the helper function like this?

Reply:
Yes, we can. I just thought it would be nicer not to call the length function in every call.

Reply:
Ah you are right, my change is making the `trace_to_string` function O(n^2)! Looks like I'm becoming too used to write functional code for proofs and not for speed :)

Maybe just add a comment above `trace_to_string_helper` explaining this is enough then.